src/ql/pass/map/qubits/map/detail/alter.cc

```diff
@@ -1,264 +1,89 @@
-        // rightopi is the index in total that holds the qubit that becomes the right operand of the gate
-        // rightopi == leftopi + 1
-        if (platform->topology->is_inter_core_hop(total[leftopi], total[rightopi])) {
-            // an inter-core hop cannot execute a two-qubit gate, so is not a valid alternative
-            // QL_DOUT("... skip inter-core hop from qubit=" << total[leftopi] << " to qubit=" << total[rightopi]);
-            continue;
+utils::List<Alter> Alter::create_from_path(const ir::PlatformRef &platform, const ir::BlockBaseRef &block, const OptionsRef &options, ir::CustomInstructionRef gate, std::list<utils::UInt> path) {
+    utils::List<Alter> result;
+
+    std::shared_ptr<std::list<utils::UInt>> shared_path(new std::list<utils::UInt>(path));
+
+    QL_ASSERT (shared_path->size() >= 2);
+    
+    auto leftOpIt = shared_path->begin();
+    auto rightOpIt = std::prev(std::prev(shared_path->rend()));
+    while (leftOpIt != std::prev(shared_path->end())) {
+        QL_ASSERT(*std::next(leftOpIt) == *rightOpIt);
+        QL_ASSERT(platform->topology->get_distance(*leftOpIt, *rightOpIt) == 1);
+        
+        // An inter-core hop cannot execute a two-qubit gate, so is not a valid alternative.
+        if (!platform->topology->is_inter_core_hop(*leftOpIt, *rightOpIt)) {
+            result.push_back(Alter(platform, block, options, gate, shared_path, leftOpIt, rightOpIt));
         }
 
-        Alter    na = *this;      // na is local copy of the current path, including total
-        // na = *this;            // na is local copy of the current path, including total
-        // na.DPRINT("... copy of current alter");
-
-        // fromSource will contain the path with qubits at indices 0 to leftopi
-        // fromTarget will contain the path with qubits at indices rightopi to length-1, reversed
-        //      reversal of fromTarget is done since swaps need to be generated starting at the target
-        utils::UInt fromi, toi;
-
-        na.from_source.resize(leftopi + 1);
-        // QL_DOUT("... fromSource size=" << na.fromSource.size());
-        for (fromi = 0, toi = 0; fromi <= leftopi; fromi++, toi++) {
-            // QL_DOUT("... fromSource: fromi=" << fromi << " toi=" << toi);
-            na.from_source[toi] = na.total[fromi];
+        if (rightOpIt != shared_path->rbegin()) {
+            rightOpIt = std::prev(rightOpIt);
         }
-
-        na.from_target.resize(length - leftopi - 1);
-        // QL_DOUT("... fromTarget size=" << na.fromTarget.size());
-        for (fromi = length-1, toi = 0; fromi > leftopi; fromi--, toi++) {
-            // QL_DOUT("... fromTarget: fromi=" << fromi << " toi=" << toi);
-            na.from_target[toi] = na.total[fromi];
-        }
-
-        // na.DPRINT("... copy of alter after split");
-        result.push_back(na);
-        // QL_DOUT("... added to result list");
-        // DPRINT("... current alter after split");
+        leftOpIt = std::next(leftOpIt);
```

Comment:
The full, unsplit path is stored (and shared) in every resulting alter (so doesn't have to be split per-se), and the partial paths information is stored as 2 pointers to the said full path. Those don't need calculations, since they come directly from iterating the full unsplit path.

Reply:
#479